sdk/python/lib/pulumi/resource.py

```diff
@@ -296,7 +296,7 @@ def __init__(self,
                  version: Optional[str] = None,
                  aliases: Optional[List['Input[Union[str, Alias]]']] = None,
                  additional_secret_outputs: Optional[List[str]] = None,
-                 id: Optional[str] = None,
+                 id: Optional['Input[str]'] = None,
```

Comment:
only actual changes in this file.  this all bottoms out on:

```py
            # Resolve the ID that we were given. Note that we are explicitly discarding the list of
            # dependencies returned to us from "serialize_property" (the second argument). This is
            # because a "read" resource does not actually have any dependencies at all in the cloud
            # provider sense, because a read resource already exists. We do not need to track this
            # dependency.
            resolved_id = await rpc.serialize_property(opts.id, [])
```

where `serialize_property` has no problem handling inputs.

Reply:
Note: with non-type-checkin python, there is nothing we need to change, since `id` works fine as any sort of `Input[str]`.  However, with type-checking python, callers downstream are forced into passing a string along.